File: Session12/17. Letter Combinations of a Phone Number.java
class Solution {
    List<String> ans=new ArrayList<>();
    String[] map = new String[]{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    public List<String> letterCombinations(String digits) {
        int n=digits.length();
        if(n==0)
        {
            return ans;
        }
        letterCombinations_rec(digits,0,n,new StringBuilder());
        return ans;
    }
    void letterCombinations_rec(String digits,int start,int n,StringBuilder temp)
    {
        if(start>=n)
        {
            ans.add(temp.toString());
            return;
        }
        int d=digits.charAt(start)-'0';
        for(char c:map[d].toCharArray())
        {
            temp.append(c);
            letterCombinations_rec(digits,start+1,n,temp);
         //   temp.delete(curr_ind,curr_ind+1);
            temp.setLength(temp.length()-1);
        }
    }
}

File: Session12/17. Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        NumMap={"1":"","2":"abc","3":"def","4":"ghi","5":"jkl","6":"mno","7":"pqrs","8":"tuv", "9":"wxyz"}
        ans=[]
        temp=[]
        n=len(digits)
        if n==0:
            return ans
        
        def helper(pos):
            if pos>=n:
                ans.append("".join(temp))
                return
            
            for c in NumMap[digits[pos]]:
                temp.append(c)
                helper(pos+1)
                temp.pop(-1)
        
        helper(0)
        return ans
            
            
            